### PR TITLE
change(state): Put chain data into an inner struct

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

The non-finalized [`Chain`](https://github.com/zcashfoundation/zebra/blob/7348d080d6dd0bb2e7153569ce4158d86861524f/zebra-state/src/service/non_finalized_state/chain.rs#L43) contains many fields that we need to manually compare in [`eq_internal_state`](https://github.com/zcashfoundation/zebra/blob/7348d080d6dd0bb2e7153569ce4158d86861524f/zebra-state/src/service/non_finalized_state/chain.rs#L265) when comparing two chains. We can't use regular `PartialEq` because we already use that for comparing only the partial cumulative work between chains: https://github.com/zcashfoundation/zebra/blob/7348d080d6dd0bb2e7153569ce4158d86861524f/zebra-state/src/service/non_finalized_state/chain.rs#L2189.

We recently added new fields into `Chain`, and we forgot to compare them in `eq_internal_state`, which was a bug that is likely to happen again.

## Solution

- Put the fields used in `eq_internal_state` into an inner struct, and derive and use `PartialEq` on this struct in `eq_internal_state`.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

**Comment by upbqdn** [comment]
I labeled this PR as a draft because wrapping the fields of `Chain` into an inner struct causes ~ 290 errors. I would like approval of the current code before I fix all the spots where we access the fields. The fixes will be trivial but tedious, so I would like to avoid them if they wouldn't be right anyway.

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 1]
```diff
@@ -48,6 +48,29 @@ pub struct Chain {
...
+    pub(super) last_fork_height: Option<Height>,
+    // # Note
+    //
+    // Most diagnostics are implemented on the NonFinalizedState, rather than each chain.
+    // Some diagnostics only use the best chain, and others need to modify the Chain state,
+    // but that's difficult with `Arc<Chain>`s.
+}
+
+/// The internal state of [`Chain`].
+#[derive(PartialEq, Clone, Debug)]
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs]: `Eq` is also guaranteed for this type, as long as all the types are `Eq` (which is checked by the derive).

https://doc.rust-lang.org/std/cmp/trait.Eq.html

```suggestion
#[derive(Clone, Debug, PartialEq, Eq)]
```

**Comment by teor2345** [comment]
> I labeled this PR as a draft because wrapping the fields of `Chain` into an inner struct causes ~ 290 errors. I would like approval of the current code before I fix all the spots where we access the fields. The fixes will be trivial but tedious, so I would like to avoid them if they wouldn't be right anyway.

The code looks fine to me.

If you `impl Deref for Chain { Target: ChainInner ... }` then most of those errors will go away, and we can see if the tests pass. You might also need to `impl DerefMut` for mutable accesses.

https://doc.rust-lang.org/std/ops/trait.Deref.html

Then we can do the mass replace in a later PR, because it's easier to review that way. (And the mass replace is likely to conflict with #7463 in this sprint.)

**Review by teor2345 — APPROVED** [review]

**Comment by teor2345** [comment]
Failed due to issue #7659, I'll restart when it's finished.

**Pull request merged**